File: ios/Runner/PlatformChannels/CaffeinePlatformChannel.swift
//
//  CaffeinePlatformChannel.swift
//  Runner
//
//  Created by Vadim Turko on 2025-05-11.
//
import Flutter

public class CaffeinePlatformChannel: NSObject {
    let methodChannel: FlutterMethodChannel
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(
            name: "com.vodemn.lightmeter.CaffeinePlatformChannel.MethodChannel",
            binaryMessenger: binaryMessenger
        )
        super.init()
        methodChannel.setMethodCallHandler({
          (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            switch call.method {
                case "isKeepScreenOn":
                    result(UIApplication.shared.isIdleTimerDisabled)
                case "setKeepScreenOn":
                    guard let keepOn = call.arguments as? Bool else {
                        result(FlutterError(code: "invalid arguments", message: "Argument should be of type Bool for 'setKeepScreenOn' call", details: nil))
                        return
                    }
                    UIApplication.shared.isIdleTimerDisabled = keepOn
                    result(true)
                default:
                    result(FlutterMethodNotImplemented)
            }
        })
        
    }
}

File: android/app/src/main/kotlin/com/vodemn/lightmeter/PlatformChannels/CaffeinePlatformChannel.kt
package com.vodemn.lightmeter.PlatformChannels

import android.view.Window
import android.view.WindowManager
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/** CaffeinePlatformChannel */
class CaffeinePlatformChannel {
    private lateinit var channel: MethodChannel

    fun onAttachedToEngine(binaryMessenger: BinaryMessenger, window: Window) {
        channel = MethodChannel(
            binaryMessenger,
            "com.vodemn.lightmeter.CaffeinePlatformChannel.MethodChannel"
        )
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "isKeepScreenOn" -> result.success((window.attributes.flags and WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON) != 0)
                "setKeepScreenOn" -> {
                    if (call.arguments !is Boolean) {
                        result.error(
                            "invalid args",
                            "Argument should be of type Bool for 'setKeepScreenOn' call",
                            null
                        )
                    } else {
                        if (call.arguments as Boolean) window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                        else window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                        result.success(true)
                    }
                }

                else -> result.notImplemented()
            }
        }
    }

    fun onDestroy() {
        channel.setMethodCallHandler(null)
    }
}
